gpt_index/indices/base.py

```diff
@@ -209,104 +205,15 @@ def query_context(self) -> Dict[str, Any]:
         """
         return {}
 
-    def _preprocess_query(self, mode: QueryMode, query_kwargs: Dict) -> None:
-        """Preprocess query.
-
-        This allows subclasses to pass in additional query kwargs
-        to query, for instance arguments that are shared between the
-        index and the query class. By default, this does nothing.
-        This also allows subclasses to do validation.
-
-        """
+    @abstractmethod
+    def as_retriever(self, **kwargs: Any) -> BaseRetriever:
```

Comment:
so, an index class can contain a retriever, but a retriever can also contain an index class? 

would prefer to avoid circular deps if possible - it makes it hard to implement future retrievers (i saw the lazy imports in the __init__ function for some of these retriever classes)

Per our convo I'm actually ok having dependence graph like this: query engine -> retriever -> index class (meaning we remove as_retriever and as_query_engine)

Reply:
to amend the statement above, some more comments:
- i wonder if we can decouple index construction from the index "state". E.g. list index state is index struct, service context, and maybe some additional hyperparameters. But we'd have an "IndexBuilder" class for each index.
- we could consider having an `IndexManager` class for each index, that implements as_retriever, and as_query_engine, and also an `get_index_builder` function perhaps. That helps preserve the high-level API so users don't have to lookup individual classes

Reply:
i'm assuming you did this to make the composable graph implementation a bit easier

Reply:
a simple thing you could do is just put the lazy imports in the `as_retriever` method instead of within each retrieval class (lazy import of the retriever class). or you could create a `get_default_retriever` class

Reply:
revised to put the lazy imports in the as_retriever method